src/protocol/ipa/mod.rs

```diff
@@ -0,0 +1,200 @@
+use std::iter::{repeat, zip};
+
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        attribution::{
+            accumulate_credit::accumulate_credit, credit_capping::credit_capping,
+            AttributionInputRow,
+        },
+        context::Context,
+        RecordId,
+    },
+    secret_sharing::{Replicated, XorReplicated},
+};
+use futures::future::try_join_all;
+
+use super::context::SemiHonestContext;
+use super::modulus_conversion::{convert_all_bits, convert_all_bits_local};
+use super::sort::generate_permutation::generate_permutation;
+use crate::protocol::boolean::bitwise_equal::bitwise_equal;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    ModulusConversionForMatchKeys,
+    GenSortPermutationFromMatchKeys,
+    ApplySortPermutation,
+    ComputeHelperBits,
+    AccumulateCredit,
+    PerformUserCapping,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::ModulusConversionForMatchKeys => "mod_conv_match_key",
+            Self::GenSortPermutationFromMatchKeys => "gen_sort_permutation_from_match_keys",
+            Self::ApplySortPermutation => "apply_sort_permutation",
+            Self::ComputeHelperBits => "compute_helper_bits",
+            Self::AccumulateCredit => "accumulate_credit",
+            Self::PerformUserCapping => "user_capping",
+        }
+    }
+}
+
+/// # Errors
+/// Propagates errors from multiplications
+#[allow(dead_code)]
+pub async fn ipa<F>(
+    ctx: SemiHonestContext<'_, F>,
+    mk_shares: &[XorReplicated],
+    num_bits: u32,
+    other_inputs: Vec<Vec<Replicated<F>>>,
+    per_user_credit_cap: u32,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    F: Field,
+{
+    debug_assert_eq!(mk_shares.len(), other_inputs.len());
+
+    let local_lists = convert_all_bits_local(ctx.role(), mk_shares, num_bits);
+    let converted_shares = convert_all_bits(
+        &ctx.narrow(&Step::ModulusConversionForMatchKeys),
+        &local_lists,
+    )
+    .await
+    .unwrap();
+    let sort_permutation = generate_permutation(
+        ctx.narrow(&Step::GenSortPermutationFromMatchKeys),
+        &converted_shares,
+        num_bits,
+    )
+    .await
+    .unwrap();
+
+    let num_bits: usize = usize::try_from(num_bits).unwrap();
+
+    let combined_match_keys_and_sidecar_data = other_inputs
+        .into_iter()
+        .enumerate()
+        .map(|(row_index, mut sidecar_data)| {
+            for i in 0..num_bits {
+                sidecar_data.push(converted_shares[i][row_index].clone());
+            }
```

Comment:
Unfortunately, `converted_shares` is _transposed_. So in this case it's a Vec of length 20, each containing a Vec of length 5. 

If anyone knows a better way of doing this transpose, please let me know.